File: src/0680_valid-palindrome-ii.c
//
/* Created by rock on 5/19/20.
*/
#include "leetcode_functions.h"

bool valid(char *s, int l, int r) {
    while (l < r ) {
        if (s[l] != s[r]) {
            return false;
        }
        l++;
        r--;
    }
    return true;
}


bool validPalindrome(char *s) {
    int l = 0;
    int r = strlen(s)-1;
    while (l < r) {
        if (s[l] != s[r]) {
            return valid(s, l, r - 1) || valid(s, l + 1, r);
        }
        l++;
        r--;
    }
    return true;


}

File: python/_0501_1000/0680_valid-palindrome-ii.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Author        : Rock Wayne 
# @Created       : 2020-05-06 08:00:00
# @Last Modified : 2020-05-06 08:00:00
# @Mail          : lostlorder@gmail.com
# @Version       : alpha-1.0

"""
# 给定一个非空字符串 s，最多删除一个字符。判断是否能成为回文字符串。 
# 
#  示例 1: 
# 
#  
# 输入: "aba"
# 输出: True
#  
# 
#  示例 2: 
# 
#  
# 输入: "abca"
# 输出: True
# 解释: 你可以删除c字符。
#  
# 
#  注意: 
# 
#  
#  字符串只包含从 a-z 的小写字母。字符串的最大长度是50000。 
#  
#  Related Topics 字符串

"""
import pytest


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:

    def validPalindrome(self, s: str) -> bool:
        def valid(left, right):
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True

        l, r = 0, len(s) - 1
        while l < r:
            if s[l] != s[r]:
                return valid(l, r - 1) or valid(l + 1, r)
            l, r = l + 1, r - 1

        return True


# leetcode submit region end(Prohibit modification and deletion)

@pytest.mark.parametrize("args,expected", [
    ("aba", True),
    ("abca", True),
])
def test_solutions(args, expected):
    assert Solution().validPalindrome(args) == expected


if __name__ == '__main__':
    pytest.main(["-q", "--color=yes", "--capture=no", __file__])
